### DISCUSSION TITLE
Context Switching이 무엇인지 설명하고 과정을 나열해주세요.7/12

### DISCUSSION THREAD

**Discussion by SionBackEnd** [Team-Psyduck/tech-sharing#69] [General]
오늘의 알고리즘 
[백준 - 리모컨](https://www.acmicpc.net/problem/1107)

**Comment by SionBackEnd** [comment]
Context Switching이란 사용자 프로세스 A에서 사용자 프로세스 B로 넘어가는 상황을 문맥 교환이라고 한다. 

Context Switching이 일어나면 사용자 프로세스 A의 작업 정보가 OS 내부 A의 PCB에 저장이 되고, OS는 B의 PCB를 확인하여 이전 작업 정보를 불러온다. 

주의할 점은 A 작업 도중에 시스템 콜을 호출하여 커널모드로 전환되었다가 다시 A의 작업으로 전환되는 것은 Context Switching이라고 하지 않는다. 
오직 다른 프로세스로 전환되는 것만을 Context Switching이라고 한다.

**Reply by kkg5** [reply]
PCB는 무엇이고 어떤 데이터들로 이루어져 있나요?
그리고 PCB는 어디에 저장되나요?

**Reply by SionBackEnd** [reply]
PCB는 Process Control Block의 약자로 CPU를 얼마나 사용했는지, 메모리를 얼마나 사용하고 있는지 확인이 가능한 자료구조입니다. 
PCB는 각 프로세스 마다 하나씩 생성되며 운영체제 data영역에 각각 생성됩니다. 
PCB에는 해당 프로세스의 Stack, Heap, ProgramCounter, Register등의 정보가 저장되며, 운영체제만이 접근 가능합니다. 
@kkg5

**Comment by eatingbug** [comment]
Context Switching 이란 하나의 프로세스가 CPU를 사용중인 상태에서 다른 프로세스가 CPU 를 사용하도록 하기 위해, 이전의 프로세스의 상태(문맥) 을 보관하고 새로 작업할 프로세스의 상태(문맥)를 적재하는 작업을 말한다.

과정은 다음 그림과 같다.
![image](https://github.com/Team-Psyduck/tech-sharing/assets/107091097/bb4aded1-7b7a-42b9-b3cd-0fa3e8369bce)
